cmake/Boost.cmake

```diff
@@ -0,0 +1,95 @@
+macro(p4c_obtain_boost)
+  option(
+    P4C_USE_PREINSTALLED_BOOST
+    "Look for a preinstalled version of Boost in the system instead of installing the library using FetchContent."
+    OFF
+  )
+
+  set(P4C_BOOST_VERSION "1.85.0")
+
+  # If P4C_USE_PREINSTALLED_BOOST is ON just try to find a preinstalled version of Boost.
+  if(P4C_USE_PREINSTALLED_BOOST)
+    if(NOT BUILD_SHARED_LIBS)
+      # Link Boost statically
+      # See https://cmake.org/cmake/help/latest/module/FindBoost.html for details
+      set(Boost_USE_STATIC_LIBS ON)
+      set(Boost_USE_STATIC_RUNTIME OFF)
+    endif()
+
+    # The boost graph headers are optional and only required by the graphs back end.
+    find_package (Boost QUIET COMPONENTS graph)
+    if (Boost_FOUND)
+      set (HAVE_LIBBOOST_GRAPH 1)
+    else ()
+      message (WARNING "Boost graph headers not found, will not build 'graphs' backend")
+    endif ()
+    find_package (Boost REQUIRED COMPONENTS format multiprecision)
+    include_directories(BEFORE SYSTEM ${Boost_INCLUDE_DIRS})
+
+  else()
+    message(STATUS "Fetching Boost version ${P4C_BOOST_VERSION} for P4C...")
+    # Print out download state while setting up Boost.
+    set(FETCHCONTENT_QUIET_PREV ${FETCHCONTENT_QUIET})
+    set(FETCHCONTENT_QUIET OFF)
+    # Unity builds do not work for Abseil...
+    set(CMAKE_UNITY_BUILD_PREV ${CMAKE_UNITY_BUILD})
+    set(CMAKE_UNITY_BUILD OFF)
+
+
+    # Add boost library sources.
+    set(BOOST_INCLUDE_LIBRARIES graph format multiprecision iostreams)
```

Comment:
This specifies what libs are built, right? We should add a mechanism for downstreams to override it as they may depend on boost too, including some libs P4C does not use. It might be enough to check if the `BOOST_INLUCDE_LIBRARIES` is set, or to list-append to the variable. Also please add a comment for this.

Reply:
I added CMake option for this.